### ISSUE TITLE
Menu items roles may not be executed on the correct web contents

### ISSUE THREAD

**Issue by mickylad** [electron/electron#6811]
* Electron version: 1.2.8
* Operating system: Windows

Hi,

I've been trying to get a context menu containing some [predefined web contents roles](https://github.com/electron/electron/blob/master/lib/browser/api/menu-item-roles.js#L17) working for an application where we have multiple windows simultaneously. I've noticed that when a context menu is opened for a window that was created later than others, the role commands don't seem to execute when clicked. However, if don't use a role and instead define a click handler that calls the command on the web contents manually, it behaves normally.

I've been doing some digging and it seems that the reason is because, for a role, [`electron.webContents.getFocusedWebContents`](https://github.com/electron/electron/blob/master/lib/browser/api/web-contents.js#L257) is used to determine which web contents to call the command on (see [menu.js](https://github.com/electron/electron/blob/master/lib/browser/api/menu.js#L121)).

This function iterates through all existing web contents, then returns the first in the collection that is in a `webview`, otherwise the first that returns `true` from `contents.isFocused()`. However, it seems to be entirely possible that there are web contents other than the one that the context menu was raised for, who think they are focused (or perhaps are within a `webview`), and who were created earlier and therefore appear earlier in the collection.

This means there are situations when the 'wrong' web contents is returned from `getFocusedWebContents`, and therefore the role command is called on that 'wrong' web contents, which leads to the unexpected behaviour described above.

Is this a bug? And if so, is there a fix and/or workaround for this? I'm happy to contribute a PR if needed.

**Comment by lukeapage** [comment]
related: https://github.com/electron/electron/issues/6203
also notes that isFocused() sometimes returns true incorrectly...

**Comment by kevinsawicki** [comment]
> Is this a bug? And if so, is there a fix and/or workaround for this? I'm happy to contribute a PR if needed.

Definitely sounds like a bug to me, are you able to provide a small app that reproduces this?

**Comment by mickylad** [comment]
Sure, I'll try to get something up tomorrow.

**Comment by mickylad** [comment]
Ok, I've recreated this here: [electron-menu-role-bug.zip](https://github.com/electron/electron/files/414951/electron-menu-role-bug.zip)

I'm creating two browser windows with `show: false`, loading the same URL for both, then showing the second (but not the first) in response to the'ready-to-show' event. The page they load has an input with a context menu for cut/copy/paste roles.

Just do `npm install` then `./node_modules/.bin/electron .` as per the [quick-start guide](https://github.com/electron/electron/blob/master/docs/tutorial/quick-start.md).

If you then try to execute some context menu actions on the input of the visible (second) window, nothing will happen. This is because the first window's web contents is being pulled up as the 'focused web contents', and the actions are being executed against that. I've added some logs to try to demonstrate that.

This doesn't occur if the windows are created without `show: false`, so it looks like the root cause may be that the hidden window doesn't toggle it's focused state properly. I'm really not sure what the expected behaviour of that is though.
